kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/InBandDekCache.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption;
+
+import java.nio.ByteBuffer;
+import java.security.SecureRandom;
+import java.util.concurrent.CompletionStage;
+
+import io.kroxylicious.kms.service.De;
+import io.kroxylicious.kms.service.Kms;
+import io.kroxylicious.kms.service.Ser;
+
+class InBandDekCache<K, E> implements DekCache<K, E> {
+
+    private final Kms<K, E> kms;
+    private final Ser<K> kekIdSerializer;
+    private final De<K> kekIdDeserializer;
+    private final Ser<E> edekSerializer;
+    private final De<E> edekDeserializer;
+
+    InBandDekCache(Kms<K, E> kms) {
+        this.kms = kms;
+        this.edekSerializer = kms.edekSerializer();
+        this.kekIdSerializer = kms.keyIdSerializer();
+        this.edekDeserializer = kms.edekDeserializer();
+        this.kekIdDeserializer = kms.keyIdDeserializer();
+    }
+
+    @Override
+    public CompletionStage<DekContext<K>> forKekId(K kekId) {
+        return kms.generateDekPair(kekId)
```

Comment:
Should we caching the generated DekContext then returning it for subsequent calls (so that we avoid the calling the provider).  

We'll also need the DekContext to somehow expire after n encrypts / passage of time, so that the next call to forKekId gets a new context backed by a fresh pair.